docs/proposals/26-workload-identity.md

```diff
@@ -0,0 +1,514 @@
+except for representing Gardener workloads before an external entity,
+eliminating the security burden to manage and store static credentials. Static
+credentials can expire or get accidentally invalidated, which will cause
+reconciliation flows to fail preventing delivery of updates, fixes, and
+improvements. This risk is better to be managed in an automated way by Gardener
+itself.
+
+The JSON Web Tokens are ephemeral and not stored anywhere by the issuer, a
+feature that Gardener can benefit of as well because it will not have to store
+the credentials of Gardener users.
+
+### Goals
+
+- Manage shoot clusters without credentials provided by the user.
+- Replace static credentials for gardener system components, e.g. DNS and backup
+  controllers.
+- Rotate credentials regularly. Rotating the token signing keys, will
+  effectively invalidate all previously issued tokens.
+- Offload Gardener users from the burden to store and manage static credentials
+  for their accounts.
+
+### Non-Goals
+
+- Register gardener as trusted identity in the shoot clusters.
+- The tokens to be usable for authentication with the Gardener API.
+- Compatibility with gardenctl integration with cloud provider CLIs.
+- OIDC compliance - just as Kubernetes, Gardener goal is not to have full OIDC
+  compliance, but to implement the bare minimum for OIDC compatible trust
+  federation.
+
+## Proposal
+
+In short, Gardener API server will generate JWTs on request by gardenlet.
+Gardenlet will ensure that the tokens can be consumed conveniently by the
+various components in the seed clusters, for example will write them in secrets.
+Service provider extensions will be responsible for making adjustments so that
+the token is consumable by the service SDK. For shoot clusters, it would be
+natural that the `cloudprovider` secret is reused as storage target for the
+token.
+
+Gardenlet will take care to refresh the token regularly and on time so that the
+target storage always contains a valid token. As an example a token can be
+refreshed when it reaches 80% of its lifetime. Of course it will all depend on
+the validity duration of such tokens. This is why these parameters will be
+configurable.
+
+The OIDC metadata discovery documents will be served in such network segment,
+e.g. public internet, so that service providers can be configured to trust
+Gardener as an OIDC compatible token issuer.
+
+### API Changes
+
+A new resource `WorkloadIdentity` in `security.gardener.cloud` API Group will be
+implemented. It will specify different characteristics of the JWT, like the
+value for the `aud` claim. The `WorkloadIdentity` resource will allow the token
+duration to be set by Gardener users so they can use tokens with shorter or
+longer validity compared to the default one. This duration will be ensured to be
+between certain limits of minimal and maximal validity, in order to avoid
+frequent token renewals as well as tokens with too long validity.
```

Comment:
>There might be compliance requirements about the validity of these tokens and global value might not be suitable for all cases, therefore it is exposed as option configurable by the users. 
There always will be a minimum validity defined (as already mentioned). The minimum validity will be the one that will serve properly all Gardener processes. I don't see a scenario where Gardener will benefit from a consumer defined longer validity.

Reply:
Gardener API and Gardenlet are provider independent components, hence they cannot bring any provider specific implementations. If the duration needs to vary for each service provider, then the provider extension admission controller needs to step in and adjust the duration, and these admission controllers are expected to make such adjustments on API objects. The `duration` field is set to be optional, which means that the default duration set by Gardener administrator will be used, unless the Gardener user decides to set own duration overwriting the default one. 

I think having such feature will not take much effort to be implemented and it should be relatively simple logic, so the change something to go wrong is low. However, if you really insist the token duration configuration must not be exposed to Gardener users, then we must drop also the minimal and maximal values as they will serve no need. At the end, there will be a Gardener API server configuration setting globally the duration of each and every workload identity token. 

PS. If needed, later we can make the duration configurable by the users within certain range.

Reply:
>  hence they cannot bring any provider specific implementations

My guess is that we will not need any provider specific implementations.

>  If the duration needs to vary for each service provider, then the provider extension admission controller needs to step in and adjust the duration, and these admission controllers are expected to make such adjustments on API objects.

Can't the extension/workload that requests the token through secret just specify the requested duration there?

I think that we can include the duration argument in the `TokenRequest` only. Then the GAPI will have min, max, default (default can equal min and thus not be provided as a parameter) and will ensure that all issued tokens are in the said range. If the requestor (gardenlet or other) requests a token outside the range then the token will be returned with the min/max configured validity depending on the requested validity. If the requestor does not specify a duration then the token can have minimum validity. WDYT?

Reply:
Yes @dimityrmirchev, this is what I had in mind with my original comment.

> If the duration needs to vary for each service provider, then the provider extension admission controller needs to step in and adjust the duration, and these admission controllers are expected to make such adjustments on API objects.

Is this really the case? This makes everything a bit more complex and if we can avoid it, we probably should?

Reply:
With @rfranzke, @timuthy and @dimityrmirchev discussed this topic and we decided to:
- drop the `duration` field from the `WorkloadIdentity` resource, if needed - it can be added later 
- add optional `duration` field in the `TokenRequest` API to let various clients to overwrite the default duration (e.g. `gardenctl` should use as short as possible duration). 
- Gardener API server will have configuration for min, default and max durations, default will be used when no duration is set in the TokenRequest. 
- As of now Gardenlet will not set explicitly duration via the TokenRequest API, effectively will be using the default duration. If needed, such configuration can be added later.

Reply:
The discussed changes are implemented with 8a967c135081b